File: packages/totem-sdk/packages/omnia/rust/src/capacity.rs
use crate::types::CapacityAssessment;

pub const WOTS_CAPACITY_TOTAL: u32 = 4096;
pub const CAPACITY_WARNING_APPROACHING: u32 = 3072; // 75%
pub const CAPACITY_WARNING_CRITICAL: u32 = 3686; // 90%
pub const CAPACITY_NEAR_EXHAUSTION: u32 = 3891; // 95%

pub fn assess_capacity(used: u32) -> Result<CapacityAssessment, String> {
    if used >= WOTS_CAPACITY_TOTAL {
        return Err(format!(
            "Channel WOTS capacity exhausted: {}/{} signing slots used",
            used, WOTS_CAPACITY_TOTAL
        ));
    }
    if used >= CAPACITY_NEAR_EXHAUSTION {
        return Ok(CapacityAssessment {
            warning: Some("critical".to_string()),
            near_exhaustion: true,
        });
    }
    if used >= CAPACITY_WARNING_CRITICAL {
        return Ok(CapacityAssessment {
            warning: Some("critical".to_string()),
            near_exhaustion: false,
        });
    }
    if used >= CAPACITY_WARNING_APPROACHING {
        return Ok(CapacityAssessment {
            warning: Some("approaching".to_string()),
            near_exhaustion: false,
        });
    }
    Ok(CapacityAssessment {
        warning: None,
        near_exhaustion: false,
    })
}

pub fn flat_signing_index(l1: u32, l2: u32) -> u32 {
    l1 * 64 + l2
}

File: packages/omnia/src/capacity.ts
import type { CapacityWarning } from './types.js';
import { ChannelCapacityError } from './errors.js';

export const WOTS_CAPACITY_TOTAL = 4096;
export const CAPACITY_WARNING_APPROACHING = Math.floor(WOTS_CAPACITY_TOTAL * 0.75);
export const CAPACITY_WARNING_CRITICAL = Math.floor(WOTS_CAPACITY_TOTAL * 0.90);
export const CAPACITY_NEAR_EXHAUSTION = Math.floor(WOTS_CAPACITY_TOTAL * 0.95);

export function assessCapacity(used: number): { warning?: CapacityWarning; nearExhaustion: boolean } {
  if (used >= WOTS_CAPACITY_TOTAL) {
    throw new ChannelCapacityError(used, WOTS_CAPACITY_TOTAL);
  }
  if (used >= CAPACITY_NEAR_EXHAUSTION) {
    return { warning: 'critical', nearExhaustion: true };
  }
  if (used >= CAPACITY_WARNING_CRITICAL) {
    return { warning: 'critical', nearExhaustion: false };
  }
  if (used >= CAPACITY_WARNING_APPROACHING) {
    return { warning: 'approaching', nearExhaustion: false };
  }
  return { nearExhaustion: false };
}

export function flatSigningIndex(l1: number, l2: number): number {
  return l1 * 64 + l2;
}
